Fix round-trip of Unstructured.OwnerReferences

ash2k opened the pull request:
**What this PR does / why we need it**:
Previously `setOwnerReference()` was storing pointers but `extractOwnerReference()` is expecting pointer fields as plain values so it cannot read those pointers. And hence you cannot read what you've just stored.

**Which issue this PR fixes**
#46817

**Special notes for your reviewer**:
This is similar to #43346.

**Release note**:
```release-note
NONE
```

ash2k commented:
/sig api-machinery

ash2k commented:
I'm not sure whether I've added tests to the right file - there is also `k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go` with some tests for unstructured.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -248,22 +248,16 @@ func extractOwnerReference(src interface{}) metav1.OwnerReference {
 	setNestedField(ret, string(src.UID), "uid")
-	setNestedField(ret, controllerPtr, "controller")
-	setNestedField(ret, blockOwnerDeletionPtr, "blockOwnerDeletion")
+	if src.Controller != nil {
```
you need comments here explaining how the marshalling leaves nil as missing, and otherwise as a bool, so this matches the "normal" marshalling.

deads2k commented:
comment about adding a comment that needs to be addressed.

deads2k commented:
@k8s-bot ok to test

deads2k commented:
@ash2k new process requires issues for bug fixes.  Make an issue, link from your description, tag me, and I'll apply the necessary labels.

ash2k committed:
Fix round-trip of Unstructured.OwnerReferences

Previously setOwnerReference was storing pointers but
extractOwnerReference is expecting pointer fields as plain values.
Fixes #46817

ash2k commented:
@deads2k added a comment to the code, created a tracking issue #46817.

shashidharatd commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref: #46827

deads2k commented:
/lgtm

deads2k commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

deads2k commented:
must be the apimachinery tests.  approved

ash2k commented:
/kind bug
